### ISSUE TITLE
Latest requirements.txt seems to be over-specifying dependencies

### ISSUE THREAD

**Issue by ocefpaf** [pyjanitor-devs/pyjanitor#760]
This is the latest version requirements.txt file:

```
black>=19.3b0
darglint
hypothesis>=4.4.0
interrogate
ipykernel
isort>=4.3.18
jupyter_client
lxml
natsort
nbsphinx>=0.4.2
pandas-flavor
pandas-vet
pre-commit
pyspark
pytest-azurepipelines
pytest-cov
pytest>=3.4.2
scikit-learn
seaborn
setuptools>=38.5.2
sphinxcontrib-fulltoc==1.2.0
unyt
xarray
```

Those are inject in the setup.py as mandatory dependencies but a lot in there looks like docs, dev, or optional. Is pyjanitor really require all that at run time?

**Comment by ericmjl** [comment]
@ocefpaf thanks for chiming in! Yes, I needed them in `requirements.txt` to ensure that Netlify deploy docs could work properly.

It seems like this breaks the conda-forge builds, is that right? Do you have ideas on how we might be able to get around this conflict?

**Comment by ocefpaf** [comment]
Doesn't really break b/c we can add them there too but it is odd to ship all that to use pyjanitor.

One solution would be stop using the requirements.txt file and specify the deps in the pyproject.toml, going full pep517/518. (I can send a PR if you want to go that route.)

PS: I don't know how netfly works but the current setup defeats the purpose of the optional deps in setup.py and the requeriments-dev.txt. The latter is a duplicate of requeriments.txt at the moment.

**Comment by ericmjl** [comment]
> PS: I don't know how netfly works but the current setup defeats the purpose of the optional deps in setup.py and the requeriments-dev.txt. The latter is a duplicate of requeriments.txt at the moment.

Totally agree. Netlify is a bit too opinionated, but the ability to preview docs changes was too alluring for me.

> One solution would be stop using the requirements.txt file and specify the deps in the pyproject.toml, going full pep517/518. (I can send a PR if you want to go that route.)

Looks like `pyproject.toml` is the right way to go here! I would love to get your help setting this up, as I've not tried this before, so seeing the pattern at work IRL will be super educational for me.

**Comment by minchinweb** [comment]
I can also suggestion using [pip-tools](https://github.com/jazzband/pip-tools). With this set up, you would have a `requirements.in` file that specified you "base" requirements, which then is used to generate a `requirements.txt` file that you can use for CI or Netlify. It's not as flashy as pyproject.toml / poetry, but it's worked well for me for several years now.

Recent versions of pip-tools would allow you to use your `setup.py` file in place of a separate `requirements.in` file as well.

**Comment by ocefpaf** [comment]
> I can also suggestion using [pip-tools](https://github.com/jazzband/pip-tools). With this set up, you would have a `requirements.in` file that specified you "base" requirements, which then is used to generate a `requirements.txt` file that you can use for CI or Netlify. It's not as flashy as pyproject.toml / poetry, but it's worked well for me for several years now.

I don't think that pyproject.toml is flashy, in fact I'm not really a fan, but it is the community standard.

> Recent versions of pip-tools would allow you to use your `setup.py` file in place of a separate `requirements.in` file as well.

I hear great things about pip-tools but I never used it. I don't see how that setup can solve the problem here though. Setup.py would still be 'bloated'' with all the dev and docs specifications to create the requirements.txt for netfly, no? Unless we add them all in separate categories and pip-tools aggregate them in a single file. (Which can also be done with pyproject.toml.)

I'll wait for your expertise on pip-tools here before sending a PR so we can choose a more stable route that would avoid another change in the short term.

**Comment by minchinweb** [comment]
> I don't think that pyproject.toml is flashy

This was perhaps unfortunate wording on my part. What I meant to say was that pyproject.toml (or poetry? they're fairly connected in my mind) first showed up promising great things (e.g. "flash"), but many of the things I was doing in `setup.py` either weren't possible or not easily discoverable. That said, the project seems to have grown into much of its original promise, and credit to the contributors who have made it happen. That said, *pip-tools* continues to work well for me, and so I've mostly worked with pyproject.toml / poetry on other people's projects, so maybe I haven't had the right exposure to realize just how awesome it is.

> I don't see how that setup can solve the problem here though.

Completely reasonable. I've prepped a PR (#761) to show how I might do it. Hopefully this will make it easier to understand what I'm proposing. And if the decision is made to go in a completely different direction, I won't be offended in the least.

>  Unless we add them all in separate categories and pip-tools aggregate them in a single file. 

This is about what I did :-/

If there are any questions, just let me know!
